contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -6391,6 +6476,796 @@ bbf_ExecDropStmt(DropStmt *stmt)
+
+		pfree(partition_name);
+		pfree(new_partition_name);
+	}
+
+	/* Free the allocated memory. */
+	if (partition_names)
+		list_free(partition_names);
+	pfree(logical_schema_name);
+	pfree(new_hash);
+	pfree(query.data);
+}
+
+/*
+ * For Babelfish partitioned tables, non-superusers should not be permitted
+ * to attach or detach partitions from the partitioned table, and they
+ * should also be restricted from modifying the partitions from both
+ * TSQL as well as PG endpoint.
+ * 
+ * NOTE: We are only blocking operation on Babelfish partitioned tables i.e.
+ * partitioned tables created from the TSQL endpoint. Existing users who have
+ * created partitioned tables from the PostgreSQL endpoint can continue to modify, 
+ * attach, and detach partitions as usual.
+ */
+static void
+bbf_alter_handle_partitioned_table(AlterTableStmt *stmt)
+{
+	AlterTableCmd		*cmd = (AlterTableCmd *) linitial(stmt->cmds);
+	int16			dbid;
+	char			*physical_schemaname;
+	char			*logical_schemaname;
+	Form_pg_class		form;
+	Oid			nsp_oid;
+	HeapTuple		tuple;
+	bool			is_partition_table, is_partitioned_table;
+	Oid			relid = RangeVarGetRelid(stmt->relation, NoLock, true);
+
+	if (!OidIsValid(relid))
+		return;
+
+	/* Get the namespace OID and rekind type of the table. */
+	tuple = SearchSysCache1(RELOID, ObjectIdGetDatum(relid));
+
+	if (!HeapTupleIsValid(tuple)) /* Sanity check. */
+		return;
+
+	form = (Form_pg_class) GETSTRUCT(tuple);
+	is_partition_table = form->relispartition;
+	is_partitioned_table = (form->relkind == 'p');
+	nsp_oid = form->relnamespace;
+	ReleaseSysCache(tuple);
+
+	/* Proceed further only if table is a partition or partitioned table. */
+	if (!is_partition_table && !is_partitioned_table)
+		return;
+
+	/* Get the schema name from namespace OID. */
+	physical_schemaname = get_namespace_name(nsp_oid);
+
+	/* Find dbid logical schema name physical schema name. */
+	logical_schemaname = (char *) get_logical_schema_name(physical_schemaname, true);
+	if (!logical_schemaname) /* not a TSQL schema */
+	{
+		pfree(physical_schemaname);
+		return;
+	}
+
+	/* Find dbid from physical schema name for a TSQL schema. */
+	dbid = get_dbid_from_physical_schema_name(physical_schemaname, false);
+
+	/*
+	 * For babelfish partitioned table, user should not be
+	 * allowed to attach/detach of table directly.
+	 * These commands can be executed only from PG endpoint.
+	 */
+	if (is_partitioned_table && (cmd->subtype == AT_AttachPartition || 
+		cmd->subtype == AT_DetachPartition || cmd->subtype == AT_DetachPartitionFinalize))
+	{
+		if (is_bbf_partitioned_table(dbid, logical_schemaname, stmt->relation->relname))
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED), 
+					errmsg("Cannot %s babelfish partitioned table '%s'.",
+							cmd->subtype == AT_AttachPartition ? "attach partition to" : "detach partition from",
+							stmt->relation->relname)));
+	}
+	/*
+	 * For partition of babelfish partitioned table, user should not
```

Comment:
Seems convoluted to understand which is being blocked

Reply:
updated the comment.